book/listings/gobject_subclassing/2/custom_button/imp.rs

```diff
@@ -0,0 +1,37 @@
+use gtk::glib;
+use gtk::prelude::*;
+use gtk::subclass::prelude::*;
+use std::cell::RefCell;
+
+// Object holding the state
+#[derive(Default)]
+pub struct CustomButton {
+    number: RefCell<i32>,
+}
+
+// The central trait for subclassing a GObject
+#[glib::object_subclass]
+impl ObjectSubclass for CustomButton {
+    const NAME: &'static str = "MyGtkAppCustomButton";
+    type Type = super::CustomButton;
+    type ParentType = gtk::Button;
+}
+
+// Trait shared by all GObjects
+impl ObjectImpl for CustomButton {
+    fn constructed(&self, obj: &Self::Type) {
+        self.parent_constructed(obj);
+        obj.set_label(&self.number.borrow().to_string());
+    }
+}
+
+// Trait shared by all widgets
+impl WidgetImpl for CustomButton {}
+
+// Trait shared by all buttons
+impl ButtonImpl for CustomButton {
+    fn clicked(&self, button: &Self::Type) {
+        *self.number.borrow_mut() += 1;
+        button.set_label(&self.number.borrow().to_string())
```

Comment:
Wouldn't it make more sense to prefix the label passed to the constructor here, for example?

Reply:
I see your point,
Actually I never need to pass a label, apart from a single time at the subclassing chapter. So I just removed `with_label` for the other CustomButton appearances.